// payload/dist/admin/components/forms/field-types/Checkbox/Input.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "CheckboxInput", {
    enumerable: true,
    get: function() {
        return CheckboxInput;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _Check = /*#__PURE__*/ _interop_require_default(require("../../../icons/Check"));
const _Line = /*#__PURE__*/ _interop_require_default(require("../../../icons/Line"));
const _Label = /*#__PURE__*/ _interop_require_default(require("../../Label"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'checkbox-input';
const CheckboxInput = (props)=>{
    const { id, name, 'aria-label': ariaLabel, checked, inputRef, label, onToggle, partialChecked, readOnly, required, className } = props;
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            baseClass,
            className,
            (checked || partialChecked) && `${baseClass}--checked`,
            readOnly && `${baseClass}--read-only`
        ].filter(Boolean).join(' ')
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__input`
    }, /*#__PURE__*/ _react.default.createElement("input", {
        "aria-label": ariaLabel,
        defaultChecked: Boolean(checked),
        disabled: readOnly,
        id: id,
        name: name,
        onInput: onToggle,
        ref: inputRef,
        type: "checkbox"
    }), /*#__PURE__*/ _react.default.createElement("span", {
        className: `${baseClass}__icon ${!partialChecked ? 'check' : 'partial'}`
    }, !partialChecked && /*#__PURE__*/ _react.default.createElement(_Check.default, null), partialChecked && /*#__PURE__*/ _react.default.createElement(_Line.default, null))), label && /*#__PURE__*/ _react.default.createElement(_Label.default, {
        htmlFor: id,
        label: label,
        required: required
    }));
};
