// payload/dist/admin/components/forms/field-types/DateTime/Input.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DateTimeInput", {
    enumerable: true,
    get: function() {
        return DateTimeInput;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _getTranslation = require("../../../../../utilities/getTranslation");
const _DatePicker = /*#__PURE__*/ _interop_require_default(require("../../../elements/DatePicker"));
const _Error = /*#__PURE__*/ _interop_require_default(require("../../Error"));
const _FieldDescription = /*#__PURE__*/ _interop_require_default(require("../../FieldDescription"));
const _Label = /*#__PURE__*/ _interop_require_default(require("../../Label"));
const _shared = require("../shared");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'date-time-field';
const DateTimeInput = (props)=>{
    const { className, datePickerProps, description, errorMessage, label, onChange, path, placeholder, readOnly, required, showError, style, value, width } = props;
    const { i18n } = (0, _reacti18next.useTranslation)();
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            _shared.fieldBaseClass,
            baseClass,
            className,
            showError && `${baseClass}--has-error`,
            readOnly && 'read-only'
        ].filter(Boolean).join(' '),
        style: {
            ...style,
            width
        }
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__error-wrap`
    }, /*#__PURE__*/ _react.default.createElement(_Error.default, {
        message: errorMessage,
        showError: showError
    })), /*#__PURE__*/ _react.default.createElement(_Label.default, {
        htmlFor: path,
        label: label,
        required: required
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__input-wrapper`,
        id: `field-${path.replace(/\./g, '__')}`
    }, /*#__PURE__*/ _react.default.createElement(_DatePicker.default, {
        ...datePickerProps,
        onChange: onChange,
        placeholder: (0, _getTranslation.getTranslation)(placeholder, i18n),
        readOnly: readOnly,
        value: value
    })), /*#__PURE__*/ _react.default.createElement(_FieldDescription.default, {
        description: description,
        value: value
    }));
};
