// payload/dist/admin/components/views/LivePreview/Toolbar/Controls/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ToolbarControls", {
    enumerable: true,
    get: function() {
        return ToolbarControls;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _ = require("../../../..");
const _ExternalLink = require("../../../../graphics/ExternalLink");
const _context = require("../../Context/context");
const _SizeInput = require("../SizeInput");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'live-preview-toolbar-controls';
const ToolbarControls = ()=>{
    const { breakpoint, breakpoints, setBreakpoint, setPreviewWindowType, setZoom, url, zoom } = (0, _context.useLivePreviewContext)();
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: baseClass
    }, breakpoints?.length > 0 && /*#__PURE__*/ _react.default.createElement("select", {
        className: `${baseClass}__breakpoint`,
        name: "live-preview-breakpoint",
        onChange: (e)=>setBreakpoint(e.target.value),
        value: breakpoint
    }, breakpoints.map((bp)=>/*#__PURE__*/ _react.default.createElement("option", {
            key: bp.name,
            value: bp.name
        }, bp.label)), breakpoint === 'custom' && // Dynamically add this option so that it only appears when the width and height inputs are explicitly changed
    // TODO: Translate this string
    /*#__PURE__*/ _react.default.createElement("option", {
        value: "custom"
    }, "Custom")), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__device-size`
    }, /*#__PURE__*/ _react.default.createElement(_SizeInput.PreviewFrameSizeInput, {
        axis: "x"
    }), /*#__PURE__*/ _react.default.createElement("span", {
        className: `${baseClass}__size-divider`
    }, /*#__PURE__*/ _react.default.createElement(_.X, null)), /*#__PURE__*/ _react.default.createElement(_SizeInput.PreviewFrameSizeInput, {
        axis: "y"
    })), /*#__PURE__*/ _react.default.createElement("select", {
        className: `${baseClass}__zoom`,
        onChange: (e)=>setZoom(Number(e.target.value) / 100),
        value: zoom * 100
    }, /*#__PURE__*/ _react.default.createElement("option", {
        value: 50
    }, "50%"), /*#__PURE__*/ _react.default.createElement("option", {
        value: 75
    }, "75%"), /*#__PURE__*/ _react.default.createElement("option", {
        value: 100
    }, "100%"), /*#__PURE__*/ _react.default.createElement("option", {
        value: 125
    }, "125%"), /*#__PURE__*/ _react.default.createElement("option", {
        value: 150
    }, "150%"), /*#__PURE__*/ _react.default.createElement("option", {
        value: 200
    }, "200%")), /*#__PURE__*/ _react.default.createElement("a", {
        className: `${baseClass}__external`,
        href: url,
        onClick: (e)=>{
            e.preventDefault();
            setPreviewWindowType('popup');
        },
        type: "button"
    }, /*#__PURE__*/ _react.default.createElement(_ExternalLink.ExternalLinkIcon, null)));
};
