// payload/utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    withMergedProps: function() {
        return _WithMergedProps.withMergedProps;
    },
    extractTranslations: function() {
        return _extractTranslations.extractTranslations;
    },
    i18nInit: function() {
        return _init.i18nInit;
    },
    combineMerge: function() {
        return _combineMerge.combineMerge;
    },
    configToJSONSchema: function() {
        return _configToJSONSchema.configToJSONSchema;
    },
    entityToJSONSchema: function() {
        return _configToJSONSchema.entityToJSONSchema;
    },
    createArrayFromCommaDelineated: function() {
        return _createArrayFromCommaDelineated.createArrayFromCommaDelineated;
    },
    deepCopyObject: function() {
        return _deepCopyObject.deepCopyObject;
    },
    deepMerge: function() {
        return _deepMerge.deepMerge;
    },
    fieldSchemaToJSON: function() {
        return _fieldSchemaToJSON.fieldSchemaToJSON;
    },
    flattenTopLevelFields: function() {
        return _flattenTopLevelFields.default;
    },
    formatLabels: function() {
        return _formatLabels.formatLabels;
    },
    formatNames: function() {
        return _formatLabels.formatNames;
    },
    toWords: function() {
        return _formatLabels.toWords;
    },
    getIDType: function() {
        return _getIDType.getIDType;
    },
    getTranslation: function() {
        return _getTranslation.getTranslation;
    },
    isValidID: function() {
        return _isValidID.isValidID;
    }
});
const _WithMergedProps = require("./dist/admin/components/utilities/WithMergedProps");
const _extractTranslations = require("./dist/translations/extractTranslations");
const _init = require("./dist/translations/init");
const _combineMerge = require("./dist/utilities/combineMerge");
const _configToJSONSchema = require("./dist/utilities/configToJSONSchema");
const _createArrayFromCommaDelineated = require("./dist/utilities/createArrayFromCommaDelineated");
const _deepCopyObject = require("./dist/utilities/deepCopyObject");
const _deepMerge = require("./dist/utilities/deepMerge");
const _fieldSchemaToJSON = require("./dist/utilities/fieldSchemaToJSON");
const _flattenTopLevelFields = /*#__PURE__*/ _interop_require_default(require("./dist/utilities/flattenTopLevelFields"));
const _formatLabels = require("./dist/utilities/formatLabels");
const _getIDType = require("./dist/utilities/getIDType");
const _getTranslation = require("./dist/utilities/getTranslation");
const _isValidID = require("./dist/utilities/isValidID");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
